py/objmodule.c

```diff
@@ -166,6 +166,11 @@ mp_obj_t mp_obj_new_module(qstr module_name) {
     return MP_OBJ_FROM_PTR(o);
 }
 
+void mp_obj_module_unregister(qstr module_name) {
```

Comment:
Instead of adding a new function to unregister a module, would it make more sense to just not register it when it is created, but rather defer registering it until we are sure that it was successfully created?

Also, there is a comment in `mp_obj_new_module()` that says:

>      // We could error out if module already exists, but let C extensions
>      // add new members to existing modules.

Which means with the proposed unregister on error handling, we could actually end up unregistering a C module when trying to extend it if there was an error in the extension code. Not sure if that is a problem since the module wouldn't be usable anyway, but something to consider.

Reply:
> Instead of adding a new function to unregister a module, would it make more sense to just not register it when it is created, but rather defer registering it until we are sure that it was successfully created?

This could be tricky; I'm not sure but I seem to recall that for e.g. nested imports and errors therein to work correctly, the module should be present early already. On the other hand there should be enough tests which would reveal that to be the case. Code size might be smaller by not creating an extra function but just putting those 2 lines where needed.

Reply:
I added a comment to the commit message and new test to justify why we still register the module early (circular imports).  

I'm not sure what specific scenario @dlech is talking about where my change would unregister a C module.  I guess there's some way to have a built-in module written in C and then a Python module with the same name that adds things to it?  I haven't come across that yet.

Reply:
> This could be tricky

Some other ideas from the related issue: https://github.com/micropython/micropython/issues/967#issuecomment-448461369